graphql_compiler/compiler/emit_sql.py

```diff
@@ -80,60 +81,54 @@ def _traverse_and_validate_blocks(ir: IrAndMetadata) -> Iterator[BasicBlock]:
         yield block
 
 
-def _find_columns_used_outside_folds(
+def _find_columns_used(
     sql_schema_info: SQLAlchemySchemaInfo, ir: IrAndMetadata
 ) -> Dict[VertexPath, Set[str]]:
-    """For each query path outside of a fold output, find which columns are used."""
+    """For each query path, find which of its columns would be used in the resulting query."""
```

Comment:
nit: Not sure if the docstring properly tells the user that the "resulting query" is part of the IR. Maybe something like: `For each query path, find all columns used in the query described by the IR.`

Reply:
This function is a bit hard to spec. The IR only talks about fields, and the query builder interprets them as columns in complicated ways. This function is a bit ugly since it tries to guess what the query builder will need. I have two different fixes in mind:
a) Instead of using sqlalchemy selectables as the main building block for queries in this module, make a new mutable selectable class that keeps track of what columns are requested of it, and can be materialized into a sqlalchemy selectable that exposes only those columns. Then this function wouldn't need to exist.
b) Document in the spec of this function what is considered a "used column", and then make sure that:
  - all objects in `CompilationState._aliases` expose all the used columns
  - only used columns are selected from `CompilationState._aliases`

Solution (a) fixes code duplication and removes some source of bugs, but it's too complicated to implement and will likely introduce other kinds of problems. Solution (b) seems right, even though it doesn't solve the duplication problem. It just creates a chain of specs from definition of "used column" to usage of "used column" that it might make this more manageable.

I suspect properly specifying this function will reveal some bugs. For instance, it's a bit unclear whether meta fields produce "used columns", and whether using a meta field inside recurse would cause a problem. I'd do this in a separate PR though, let me know if you agree.

Reply:
I can wrap part of this in a todo that explains what "the resulting query" is

Reply:
Just wrote a clarifying todo

Reply:
Agreed that solution (a) seems too complicated. The todo helps clarify things. Thanks!